### ISSUE TITLE
[RFC] FA4 pt compile plan

### ISSUE THREAD

**Issue by drisspg** [Dao-AILab/flash-attention#2154]
# RFC: FA4 + PT2 full graph

Driss (HUMAN NOTE): I had claude write this but I came up with this design. I just did a few rounds with it to add the details I wanted.

## Summary

Make `flash_attn_func` and `flash_attn_varlen_func` fully PT2-compliant by registering the forward/backward kernels as custom ops.

**Key changes:**
- Register custom ops for forward: `flash_attn::flash_attn_fwd`, `flash_attn::flash_attn_varlen_fwd`
- Register custom ops for backward: `flash_attn::flash_attn_bwd`, `flash_attn::flash_attn_varlen_bwd`
- Use `autograd.Function` that calls these custom ops (simpler than `register_autograd`, per Richard Zou)
- Blocksparse/callable attention stays eager-only (use `flex_attention` with Flash backend for PT2 + callables)

> **PT2 + Callables:** Use `flex_attention(..., kernel_options={"BACKEND": "FLASH"})` for full `torch.compile` support with `score_mod`/`mask_mod`. See [flex_flash_attention.py](https://github.com/meta-pytorch/attention-gym/blob/main/examples/flex_flash_attention.py).

Driss: There are some much newer apis that might make this possible without flex but this is fully fleshed out and we should recommend, you need nightly PT though.

## Current vs Proposed

**Before:**
```
flash_attn_func() → FlashAttnFunc.apply() → _flash_attn_fwd()  (not a custom op)
```

**After:**
```
flash_attn_func() → FlashAttnFunc.apply() → torch.ops.flash_attn.flash_attn_fwd  (custom op!)
                                          ↳ torch.ops.flash_attn.flash_attn_bwd  (custom op!)
```

The `autograd.Function` stays, but now it calls registered custom ops instead of raw functions.

## API Summary

| Function | Callables | torch.compile | Use Case |
|----------|-----------|---------------|----------|
| `flash_attn_func()` | ❌ | ✅ | Production inference, AOTInductor |
| `flash_attn_varlen_func()` | ❌ | ✅ | Packed sequences, production |
| `flash_attn_blocksparse_func()` | ✅ | ❌ | Blocksparse, callables (eager-only) |
| `flex_attention(kernel_options={"BACKEND": "FLASH"})` | ✅ | ✅ | PT2 + callables |

## Migration

1. **No callables**: No changes needed - same API, now PT2-compliant
2. **With callables**: Use `flash_attn_blocksparse_func()` (eager) or `flex_attention` (PT2)
3. **Blocksparse**: Use `flash_attn_blocksparse_func()` with `block_sparse_tensors`

## References

- [PyTorch Custom Python Operators](https://docs.pytorch.org/tutorials/advanced/python_custom_ops.html)
- [FlexAttention + Flash Backend](https://github.com/meta-pytorch/attention-gym/blob/main/examples/flex_flash_attention.py)

---

### Architecture

```
interface.py
├── Custom Operators (kernels only, no autograd)
│   ├── flash_attn::flash_attn_fwd      + @register_fake
│   ├── flash_attn::flash_attn_bwd      + @register_fake
│   ├── flash_attn::flash_attn_varlen_fwd + @register_fake
│   └── flash_attn::flash_attn_varlen_bwd + @register_fake
│
├── Autograd Functions (call custom ops)
│   ├── FlashAttnFunc.forward()  → torch.ops.flash_attn.flash_attn_fwd
│   ├── FlashAttnFunc.backward() → torch.ops.flash_attn.flash_attn_bwd
│   └── FlashAttnVarlenFunc (same pattern)
│
├── Public API
│   ├── flash_attn_func() → FlashAttnFunc.apply()
│   └── flash_attn_varlen_func() → FlashAttnVarlenFunc.apply()
│
└── Eager-Only (blocksparse + callables)
    └── flash_attn_blocksparse_func()
```

### Custom Operators (kernels only)

Note: The internal `_flash_attn_fwd` and `_flash_attn_bwd` functions still accept `score_mod`, `mask_mod`, and `block_sparse_tensors` arguments - we just don't expose them in the custom op signatures (they remain eager-only via `flash_attn_blocksparse_func`).

```python
import torch
from torch import Tensor

@torch.library.custom_op("flash_attn::flash_attn_fwd", mutates_args=())
def flash_attn_fwd(
    q: Tensor, k: Tensor, v: Tensor,
    softmax_scale: float, causal: bool,
    window_size_left: int, window_size_right: int,
    softcap: float, deterministic: bool,
) -> tuple[Tensor, Tensor]:
    """Returns (out, lse)"""
    return _flash_attn_fwd(q, k, v, softmax_scale, causal,
        (window_size_left, window_size_right), softcap, deterministic)


@torch.library.custom_op("flash_attn::flash_attn_bwd", mutates_args=())
def flash_attn_bwd(
    dout: Tensor, q: Tensor, k: Tensor, v: Tensor,
    out: Tensor, lse: Tensor,
    softmax_scale: float, causal: bool,
    window_size_left: int, window_size_right: int,
    softcap: float, deterministic: bool,
) -> tuple[Tensor, Tensor, Tensor]:
    """Returns (dq, dk, dv)"""
    return _flash_attn_bwd(dout, q, k, v, out, lse, softmax_scale, causal,
        (window_size_left, window_size_right), softcap, deterministic)


@torch.library.custom_op("flash_attn::flash_attn_varlen_fwd", mutates_args=())
def flash_attn_varlen_fwd(
    q: Tensor, k: Tensor, v: Tensor,
    cu_seqlens_q: Tensor, cu_seqlens_k: Tensor,
    max_seqlen_q: int, max_seqlen_k: int,
    softmax_scale: float, causal: bool,
    window_size_left: int, window_size_right: int,
    softcap: float, deterministic: bool,
) -> tuple[Tensor, Tensor]:
    """Returns (out, lse)"""
    return _flash_attn_fwd(q, k, v, softmax_scale, causal,
        (window_size_left, window_size_right), softcap, deterministic,
        cu_seqlens_q=cu_seqlens_q, cu_seqlens_k=cu_seqlens_k,
        max_seqlen_q=max_seqlen_q, max_seqlen_k=max_seqlen_k)


@torch.library.custom_op("flash_attn::flash_attn_varlen_bwd", mutates_args=())
def flash_attn_varlen_bwd(
    dout: Tensor, q: Tensor, k: Tensor, v: Tensor,
    out: Tensor, lse: Tensor,
    cu_seqlens_q: Tensor, cu_seqlens_k: Tensor,
    max_seqlen_q: int, max_seqlen_k: int,
    softmax_scale: float, causal: bool,
    window_size_left: int, window_size_right: int,
    softcap: float, deterministic: bool,
) -> tuple[Tensor, Tensor, Tensor]:
    """Returns (dq, dk, dv)"""
    return _flash_attn_bwd(dout, q, k, v, out, lse, softmax_scale, causal,
        (window_size_left, window_size_right), softcap, deterministic,
        cu_seqlens_q=cu_seqlens_q, cu_seqlens_k=cu_seqlens_k,
        max_seqlen_q=max_seqlen_q, max_seqlen_k=max_seqlen_k)
```

### FakeTensor Registration

```python
@flash_attn_fwd.register_fake
def flash_attn_fwd_fake(q, k, v, softmax_scale, causal,
                        window_size_left, window_size_right, softcap, deterministic):
    batch, seqlen_q, nheads, hdim = q.shape
    hdim_v = v.shape[-1]
    return q.new_empty(batch, seqlen_q, nheads, hdim_v), \
           q.new_empty(batch, nheads, seqlen_q, dtype=torch.float32)


@flash_attn_bwd.register_fake
def flash_attn_bwd_fake(dout, q, k, v, out, lse, softmax_scale, causal,
                        window_size_left, window_size_right, softcap, deterministic):
    return torch.empty_like(q), torch.empty_like(k), torch.empty_like(v)


@flash_attn_varlen_fwd.register_fake
def flash_attn_varlen_fwd_fake(q, k, v, cu_seqlens_q, cu_seqlens_k,
                                max_seqlen_q, max_seqlen_k, softmax_scale, causal,
                                window_size_left, window_size_right, softcap, deterministic):
    total_q, nheads, hdim = q.shape
    hdim_v = v.shape[-1]
    return q.new_empty(total_q, nheads, hdim_v), \
           q.new_empty(nheads, total_q, dtype=torch.float32)


@flash_attn_varlen_bwd.register_fake
def flash_attn_varlen_bwd_fake(dout, q, k, v, out, lse, cu_seqlens_q, cu_seqlens_k,
                                max_seqlen_q, max_seqlen_k, softmax_scale, causal,
                                window_size_left, window_size_right, softcap, deterministic):
    return torch.empty_like(q), torch.empty_like(k), torch.empty_like(v)
```

### Autograd Functions (call custom ops)

```python
class FlashAttnFunc(torch.autograd.Function):
    @staticmethod
    def forward(ctx, q, k, v, softmax_scale, causal, window_size, softcap, deterministic):
        out, lse = torch.ops.flash_attn.flash_attn_fwd(
            q, k, v, softmax_scale, causal,
            window_size[0], window_size[1], softcap, deterministic)
        ctx.save_for_backward(q, k, v, out, lse)
        ctx.softmax_scale = softmax_scale
        ctx.causal = causal
        ctx.window_size = window_size
        ctx.softcap = softcap
        ctx.deterministic = deterministic
        return out, lse

    @staticmethod
    def backward(ctx, dout, dlse):
        q, k, v, out, lse = ctx.saved_tensors
        dq, dk, dv = torch.ops.flash_attn.flash_attn_bwd(
            dout, q, k, v, out, lse,
            ctx.softmax_scale, ctx.causal,
            ctx.window_size[0], ctx.window_size[1],
            ctx.softcap, ctx.deterministic)
        return dq, dk, dv, None, None, None, None, None


class FlashAttnVarlenFunc(torch.autograd.Function):
    @staticmethod
    def forward(ctx, q, k, v, cu_seqlens_q, cu_seqlens_k, max_seqlen_q, max_seqlen_k,
                softmax_scale, causal, window_size, softcap, deterministic):
        out, lse = torch.ops.flash_attn.flash_attn_varlen_fwd(
            q, k, v, cu_seqlens_q, cu_seqlens_k, max_seqlen_q, max_seqlen_k,
            softmax_scale, causal,
            window_size[0], window_size[1], softcap, deterministic)
        ctx.save_for_backward(q, k, v, out, lse, cu_seqlens_q, cu_seqlens_k)
        ctx.max_seqlen_q = max_seqlen_q
        ctx.max_seqlen_k = max_seqlen_k
        ctx.softmax_scale = softmax_scale
        ctx.causal = causal
        ctx.window_size = window_size
        ctx.softcap = softcap
        ctx.deterministic = deterministic
        return out, lse

    @staticmethod
    def backward(ctx, dout, dlse):
        q, k, v, out, lse, cu_seqlens_q, cu_seqlens_k = ctx.saved_tensors
        dq, dk, dv = torch.ops.flash_attn.flash_attn_varlen_bwd(
            dout, q, k, v, out, lse, cu_seqlens_q, cu_seqlens_k,
            ctx.max_seqlen_q, ctx.max_seqlen_k,
            ctx.softmax_scale, ctx.causal,
            ctx.window_size[0], ctx.window_size[1],
            ctx.softcap, ctx.deterministic)
        return dq, dk, dv, None, None, None, None, None, None, None, None, None


def flash_attn_func(q, k, v, softmax_scale=None, causal=False, window_size=(-1, -1),
                    softcap=0.0, deterministic=False):
    if softmax_scale is None:
        softmax_scale = q.shape[-1] ** -0.5
    return FlashAttnFunc.apply(q, k, v, softmax_scale, causal, window_size, softcap, deterministic)


def flash_attn_varlen_func(q, k, v, cu_seqlens_q, cu_seqlens_k, max_seqlen_q, max_seqlen_k,
                           softmax_scale=None, causal=False, window_size=(-1, -1),
                           softcap=0.0, deterministic=False):
    if softmax_scale is None:
        softmax_scale = q.shape[-1] ** -0.5
    return FlashAttnVarlenFunc.apply(q, k, v, cu_seqlens_q, cu_seqlens_k,
                                      max_seqlen_q, max_seqlen_k, softmax_scale,
                                      causal, window_size, softcap, deterministic)
```

### Blocksparse (Eager-Only)

```python
def flash_attn_blocksparse_func(q, k, v, block_sparse_tensors=None, softmax_scale=None,
                                 causal=False, window_size=(-1, -1), softcap=0.0,
                                 deterministic=False, score_mod=None, mask_mod=None):
    """Eager-only. Delegates to compilable path if no callables/sparsity."""
    if block_sparse_tensors is None and score_mod is None and mask_mod is None:
        return flash_attn_func(q, k, v, softmax_scale, causal, window_size, softcap, deterministic)
    # ... blocksparse implementation (eager only, not shown)
```

### Testing

```python
from torch.library import opcheck

def test_opcheck():
    sample_inputs = [(
        torch.randn(2, 128, 8, 64, device="cuda", dtype=torch.float16),
        torch.randn(2, 128, 8, 64, device="cuda", dtype=torch.float16),
        torch.randn(2, 128, 8, 64, device="cuda", dtype=torch.float16),
        0.125, False, -1, -1, 0.0, False,
    )]
    opcheck(torch.ops.flash_attn.flash_attn_fwd, sample_inputs)
    opcheck(torch.ops.flash_attn.flash_attn_bwd, ...)  # similar


def test_torch_compile_fullgraph():
    @torch.compile(fullgraph=True)
    def compiled_attn(q, k, v):
        return flash_attn_func(q, k, v)

    q = torch.randn(2, 128, 8, 64, device="cuda", dtype=torch.float16)
    k = torch.randn(2, 128, 8, 64, device="cuda", dtype=torch.float16)
    v = torch.randn(2, 128, 8, 64, device="cuda", dtype=torch.float16)
    out, lse = compiled_attn(q, k, v)
    assert out.shape == (2, 128, 8, 64)
```

**Comment by drisspg** [comment]
CC @reubenconducts

**Comment by gilfordting** [comment]
Since this plan touches the top-level interfaces: should we have `return_lse` as a parameter as well? Hopper/FA3 supports this as an option, and it's useful for e.g. Ring Attention.

**Comment by reubenconducts** [comment]
Do we want to allow block sparsity without callables? It'd be a non-trivial change to the kernels but it might be important.

**Comment by gilfordting** [comment]
This should be feasible -- it's not there rn because the `BlockSparseTensorsTorch` type is incompatible with torch.compile, but that is just a wrapper around 4 `torch.Tensor`s; if we just construct the `BlockSparseTensorsTorch` within the custom op it should work fine

**Comment by reubenconducts** [comment]
I was slightly overthinking this; we just need to pass in `mask_block_*` tensors that are identically zero and `full_block_*` tensors that indicate where to compute. The non-trivial aspect is e.g. allowing causal/local + blocksparsity.

**Comment by drisspg** [comment]
ehhh I feel like block sparsity without the callables is pretty foot gunny, unless like you mentioned we get causal/local + block sparse working so that we have some intra block masking

**Comment by vadimkantorov** [comment]
This also applies to FAv2, right?
- https://github.com/dao-AILab/flash-attention/issues/2071

Then HF could also support fullgraph with FA:
- https://github.com/huggingface/transformers/issues/42950

**Comment by vadimkantorov** [comment]
Also, if regular torch.autograd.Function is used, it would be great if it directly used the new-style with a separate setup_context method. Then these autograd functions would be supported by torch.func functions (including torch.func.grad)

**Comment by tridao** [comment]
> Also, if regular torch.autograd.Function is used, it would be great if it directly used the new-style with a separate setup_context method. Then these autograd functions would be supported by torch.func functions (including torch.func.grad)

I'm personally not a fan of the new-style. Perhaps I don't understand it yet (old man yelling at cloud?).

**Comment by zou3519** [comment]
@vadimkantorov why do you want to use torch.func.grad?

@tridao I think it's valid to say that pre-existing APIs that folks have used for years are very difficult to get people off of

**Comment by vadimkantorov** [comment]
For fullgraph compilation of getting a grad. For now, it seems it's the only option. Adding setup_context to the autograd.Funcion impls here seems very doable, mainly seems just a small refactoring separating the concerns

The other remaining thing might be fixing the functions accepting some args as only Python ints instead of also accepting scalar Tensors, this should fix it for CUDA graphs (for varlen / max_q/ max_k)

**Comment by vadimkantorov** [comment]
@tridao for setup_context refactoring example, I think it would look sth like this - just splits out `save_for_backward` calls into `setup_context` function, and might require adding more outputs to the `forward` function (so that some of these temp tensors can be used in `setup_context`):
- https://github.com/Dao-AILab/flash-attention/issues/2071#issuecomment-3672992886
